File: backend/package/yuxi/utils/question_utils.py
"""问题和选项规范化工具"""

import uuid
from typing import Any


def normalize_options(raw_options: Any) -> list[dict[str, str]]:
    """规范化选项列表"""
    if not isinstance(raw_options, list):
        return []

    options: list[dict[str, str]] = []
    for item in raw_options:
        if isinstance(item, dict):
            label = str(item.get("label") or item.get("value") or "").strip()
            value = str(item.get("value") or item.get("label") or "").strip()
        else:
            label = str(item).strip()
            value = label
        if label and value:
            options.append({"label": label, "value": value})
    return options


def normalize_questions(raw_questions: Any, default_question_id_prefix: str = "q") -> list[dict[str, Any]]:
    """规范化问题列表"""
    if not isinstance(raw_questions, list):
        return []

    questions: list[dict[str, Any]] = []
    for idx, item in enumerate(raw_questions):
        if not isinstance(item, dict):
            continue

        question = str(item.get("question") or "").strip()
        if not question:
            continue

        question_id = str(item.get("question_id") or f"{default_question_id_prefix}-{idx + 1}").strip()
        if not question_id:
            question_id = str(uuid.uuid4())

        normalized_question: dict[str, Any] = {
            "question_id": question_id,
            "question": question,
            "options": normalize_options(item.get("options")),
            "multi_select": bool(item.get("multi_select", False)),
            "allow_other": bool(item.get("allow_other", True)),
        }

        operation = item.get("operation")
        if isinstance(operation, str) and operation.strip():
            normalized_question["operation"] = operation.strip()

        questions.append(normalized_question)

    return questions


def normalize_legacy_question(raw_question: Any) -> dict[str, Any] | None:
    """规范化单个问题（兼容旧格式）"""
    if not raw_question:
        return None

    question = str(raw_question.get("question") or "").strip()
    if not question:
        return None

    question_id = str(raw_question.get("question_id") or "").strip()
    if not question_id:
        question_id = str(uuid.uuid4())

    return {
        "question_id": question_id,
        "question": question,
        "options": normalize_options(raw_question.get("options")),
        "multi_select": bool(raw_question.get("multi_select", False)),
        "allow_other": bool(raw_question.get("allow_other", True)),
    }

File: web/src/utils/questionUtils.js
/**
 * 问题和选项规范化工具
 */

const DEFAULT_OTHER_OPTION_VALUE = '__other__'

/**
 * 判断选项是否为"其他"选项
 */
export const isOtherOption = (option) => {
  if (!option || typeof option !== 'object') return false
  const label = String(option.label || '')
    .trim()
    .toLowerCase()
  const value = String(option.value || '')
    .trim()
    .toLowerCase()

  return (
    value === DEFAULT_OTHER_OPTION_VALUE ||
    value === 'other' ||
    label.includes('其他') ||
    label.includes('other')
  )
}

/**
 * 规范化选项列表
 */
export const normalizeOptions = (rawOptions) => {
  if (!Array.isArray(rawOptions)) return []

  return rawOptions
    .map((item) => {
      if (item && typeof item === 'object') {
        const label = String(item.label || item.value || '').trim()
        const value = String(item.value || item.label || '').trim()
        return label && value ? { label, value } : null
      }

      const text = String(item || '').trim()
      return text ? { label: text, value: text } : null
    })
    .filter(Boolean)
}

/**
 * 规范化问题列表
 */
export const normalizeQuestions = (rawQuestions) => {
  if (!Array.isArray(rawQuestions)) return []

  return rawQuestions
    .map((item, index) => {
      if (!item || typeof item !== 'object') return null

      const question = String(item.question || '').trim()
      if (!question) return null

      const questionId =
        String(item.questionId || item.question_id || '').trim() || `q-${index + 1}`
      const operation = String(item.operation || '').trim()
      const allowOther = Boolean(item.allowOther ?? item.allow_other ?? true)
      const baseOptions = normalizeOptions(item.options || [])
      const hasOtherOption = baseOptions.some((option) => isOtherOption(option))
      const options =
        allowOther && !hasOtherOption
          ? [...baseOptions, { label: '其他', value: DEFAULT_OTHER_OPTION_VALUE }]
          : baseOptions

      return {
        questionId,
        question,
        options,
        multiSelect: Boolean(item.multiSelect ?? item.multi_select ?? false),
        allowOther,
        operation
      }
    })
    .filter(Boolean)
}

/**
 * 从旧格式构建单个问题（向后兼容）
 */
export const buildLegacyQuestion = (chunk, interruptInfo) => {
  const question = String(chunk?.question || interruptInfo?.question || '').trim()
  if (!question) return null

  const operation = String(chunk?.operation || interruptInfo?.operation || '').trim()

  return {
    questionId: String(chunk?.question_id || interruptInfo?.question_id || '').trim() || 'q-1',
    question,
    options: normalizeOptions(chunk?.options || interruptInfo?.options || []),
    multiSelect: Boolean(chunk?.multi_select ?? interruptInfo?.multi_select ?? false),
    allowOther: Boolean(chunk?.allow_other ?? interruptInfo?.allow_other ?? true),
    operation
  }
}

export { DEFAULT_OTHER_OPTION_VALUE }
